// nilssongames-tetris/lib/board.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
function Board() {
  this.width = 200;
  this.height = 400;
  this.x = 0;
  this.y = 0;
  this.color = '#242435';
  this.grid = [[0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]];
  this.gameOver = false;

  this.draw = function (context) {
    context.fillStyle = this.color;
    context.fillRect(this.x, this.y, this.width, this.height);

    for (var row = this.grid.length - 1; row >= 0; row--) {
      var rowFilled = true;

      for (var col = this.grid[row].length - 1; col >= 0; col--) {
        if (this.grid[row][col] === 0) {
          var x = col * 20;
          var y = row * 20;
          context.fillStyle = 'rgba(0, 0, 0, 0.2)';
          context.fillRect(x + 1, y + 18, 20 - 2, 20 - 18);
          context.fillRect(x + 18, y + 1, 20 - 19, 20 - 1);
          rowFilled = false;
        }
      }

      if (rowFilled) {
        this.grid.splice(row, 1);
        this.grid.unshift([0, 0, 0, 0, 0, 0, 0, 0, 0, 0]);
      }
    }

    for (var row = this.grid.length - 1; row >= 0; row--) {
      for (var col = this.grid[row].length - 1; col >= 0; col--) {
        if (this.grid[row][col] !== 0) {
          context.fillStyle = this.grid[row][col];
          var x = col * 20;
          var y = row * 20;
          context.fillRect(x + 1, y + 1, 20 - 2, 20 - 2);
          context.fillStyle = 'rgba(255, 255, 255, 0.4)';
          context.fillRect(x + 1, y + 1, 20 - 2, 20 - 18);
          context.fillRect(x + 1, y + 1, 20 - 19, 20 - 1);
          context.fillStyle = 'rgba(0, 0, 0, 0.4)';
          context.fillRect(x + 1, y + 18, 20 - 2, 20 - 18);
          context.fillRect(x + 18, y + 1, 20 - 19, 20 - 1);
        }
      }
    }

    if (this.gameOver) {
      context.fillStyle = 'red';
      context.font = '36px Arial';
      context.textAlign = 'center';
      context.fillText('Game over', this.width / 2, this.height / 4);
    }
  };
}

exports.default = Board;
